Module 2: JS Challenge

juanescobarsv opened the pull request:
Solved the 5 exercises

![image](https://github.com/user-attachments/assets/45a67cad-085d-4356-894c-678e51be4f67)

juanescobarsv committed:
feat/Challenge1+2

Solutions for challenges 1 and 2.

juanescobarsv committed:
feat/challenge-completed

Exercises 1-4 completed, tested and passed.

juanescobarsv committed:
feat/added-explanations + fix/excercise1

Added explanations to each solution and added a validation for exercise1, making sure it's a positive number.

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -13,9 +13,20 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
+  if (typeof seconds !== "number" || seconds < 0) {
+    return "Invalid input: Please provide a non-negative number of seconds.";
+  }
+  const HH = String(Math.floor(seconds / 3600)).padStart(2, "0");
+  const MM = String(Math.floor((seconds % 3600) / 60)).padStart(2, "0");
+  const SS = String(seconds % 60).padStart(2, "0");
```
Small nit: use caps for local variables

```suggestion
  const hh = String(Math.floor(seconds / 3600)).padStart(2, "0");
  const mm = String(Math.floor((seconds % 3600) / 60)).padStart(2, "0");
  const ss = String(seconds % 60).padStart(2, "0");
```

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -13,9 +13,20 @@ Invoking "readableTime(3690)" should return "01:01:30" (HH:MM:SS)
 ***** */
 
 const readableTime = (seconds) => {
```
Looking great!

juanescobarsv replied:
TY!

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -41,9 +52,16 @@ Invoking "circularArray(2)" should return "["Island", "Japan", "Israel", "German
 
 const circularArray = (index) => {
-  // YOUR CODE HERE...
+  const validIndex = index % COUNTRY_NAMES.length;
+  const part1 = COUNTRY_NAMES.slice(validIndex);
+  const part2 = COUNTRY_NAMES.slice(0, validIndex);
+  return part1.concat(part2);
```
Consider validating the input `index`, defensive programming is a good practice but it looks great.

juanescobarsv replied:
Ok thanks, will keep in mind to start with a validation next time to avoid any possible issues/bugs!

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -70,9 +88,18 @@ The last 3 digits for the sum of powers from 1 to 10 is "317"
 ***** */
 
 const ownPower = (number, lastDigits) => {
```
Consider validating inputs (number and lastDigits) for type and bounds. What happens if negative numbers or non-integer values are passed?

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -95,9 +122,23 @@ Since 10! === 3628800 and you sum 3 + 6 + 2 + 8 + 8 + 0 + 0
 ***** */
 
 const digitSum = (n) => {
```
Can we validate the input please?

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -95,9 +122,23 @@ Since 10! === 3628800 and you sum 3 + 6 + 2 + 8 + 8 + 0 + 0
 
 const digitSum = (n) => {
-  // YOUR CODE HERE...
+  let factorial = 1n;
```
why BigInt?

juanescobarsv replied:
for precision sake, it helps with big numbers and calculations on JS

franlopz commented on JS-Algorithms/challenges.js:
```diff
@@ -118,9 +159,25 @@ Because the 12th index in the Fibonacci sequence is 144, and 144 has three digit
 
 const fibIndex = (n) => {
-  // YOUR CODE HERE...
+  let a = 1n,
+    b = 1n;
+  let index = 2;
+
+  while (String(b).length < n) {
+    const next = a + b;
+    a = b;
+    b = next;
+    index++;
+  }
+
+  return index;
```
Great job! Just a recommendation: Validate the input.

juanescobarsv committed:
Update JS-Algorithms/challenges.js

ok! Local variables shouldn't start with Caps, got it! Thanks Fran

Co-authored-by: franlopz <92971105+franlopz@users.noreply.github.com>

juanescobarsv committed:
fix/validations

Added validations to each function to ensure that the parameters are both numbers and integers to avoid any issues.

juanescobarsv commented:
Validations added to each function now! Fixes waiting for Approval
